crates/codegen/parser/runtime/src/cst.rs

```diff
@@ -46,63 +46,103 @@ impl Node {
         }
     }
 
-    pub fn create_cursor(&self, text_offset: TextIndex) -> Cursor {
+    /// Creates a [`Cursor`] that starts at the current node as the root and a given initial `text_offset`.
+    pub fn cursor_with_offset(&self, text_offset: TextIndex) -> Cursor {
         Cursor::new(self.clone(), text_offset)
     }
 
+    /// Reconstructs the original source code from the parse tree.
+    pub fn reconstruct(self) -> String {
+        match self {
+            Self::Rule(rule) => rule.reconstruct(),
+            Self::Token(token) => token.text.clone(),
+        }
+    }
+
     pub fn as_rule(&self) -> Option<&Rc<RuleNode>> {
         match self {
             Self::Rule(node) => Some(node),
             _ => None,
         }
     }
 
+    pub fn into_rule(self) -> Option<Rc<RuleNode>> {
+        match self {
+            Self::Rule(node) => Some(node),
+            _ => None,
+        }
+    }
+
     pub fn as_token(&self) -> Option<&Rc<TokenNode>> {
         match self {
             Self::Token(node) => Some(node),
             _ => None,
         }
     }
 
-    pub fn as_token_with_kind(&self, kinds: &[TokenKind]) -> Option<&Rc<TokenNode>> {
-        if let Node::Token(token_node) = self {
-            if kinds.contains(&token_node.kind) {
-                return Some(token_node);
-            }
+    pub fn into_token(self) -> Option<Rc<TokenNode>> {
+        match self {
+            Self::Token(node) => Some(node),
+            _ => None,
         }
-        return None;
+    }
+
+    pub fn as_token_with_kind(&self, kinds: &[TokenKind]) -> Option<&Rc<TokenNode>> {
+        self.as_token_matching(|token| kinds.contains(&token.kind))
     }
 
     pub fn as_token_matching<F: Fn(&Rc<TokenNode>) -> bool>(
         &self,
         predicate: F,
     ) -> Option<&Rc<TokenNode>> {
-        if let Node::Token(token_node) = self {
-            if predicate(&token_node) {
-                return Some(token_node);
-            }
+        match self {
+            Node::Token(token) if predicate(&token) => Some(token),
+            _ => None,
         }
-        return None;
     }
 
     pub fn as_rule_with_kind(&self, kinds: &[RuleKind]) -> Option<&Rc<RuleNode>> {
```

Comment:
Possibly unrelated to this PR, but asking since it is addressing API feedback:

For `as_rule_matching(...)` .. I wonder how is that different from `.as_rule().filter(...)`, given it is the same characters/effort to type?

Are we duplicating Rust `std::` methods? what new scenarios does it enable?

Same question for `as_token_matching()`.

Reply:
You're right that this is merely a convenience/"combined" method.

This API is from when, as I understand it, we wanted to have parity in the Rust/TS API. Since we decided against that recently, we can start thinking whether we want it or not on Rust side (AIUI it's more performant to keep that on TS side).

My first reaction was the same as yours, i.e `as_{rule,token}_matching` is probably redundant on the Rust side, now that we have introduced convenience `as_{rule,token}` at some point.

I do see some point in the convenience `_with_kind` functions, as `as_{rule,token}_with_kind` reads and formats better than the `as_{rule,token}().filter(|node| &[some,kinds].contains(node.kind))` chain. WDYT?

Reply:
I like these semantically rich apis that match common use cases. I'm also happy to have them in a separate trait, such as RichXXXAPI or EnhancedXXX etc, simply for the sake of having a 'core' minimal API and then a layered convenience API.

Reply:
Specifically because it reads well in the code to express clearly the higher-level intent. And I like descriptive precise names, with no concern at all about the length of names.

Reply:
> For as_rule_matching(...) .. I wonder how is that different from .as_rule().filter(...), given it is the same characters/effort to type?

Same to type, but less to understand.

Reply:
> I do see some point in the convenience `_with_kind` functions, as `as_{rule,token}_with_kind` reads and formats better than the `as_{rule,token}().filter(|node| &[some,kinds].contains(node.kind))` chain. WDYT?

> Same to type, but less to understand.

@Xanewok @AntonyBlakey

`with_kind()` makes sense. The alternative doesn't have a good Rust standard library alternative, and I imagine it is a common use case, enough to have a helper for.

But on the other hand, `as_rule_matching()` here is longer than the Rust standard library alternative (`.as_rule().filter()`), which most Rust users are already familiar with, and is used extensively in all projects, so I feel we are only replacing the language built-in API with a longer/less idiomatic one, that is not really easier to use (both accept the same kind of callback).

Reply:
Agree, I'm not so sure about the _matching variant, especially since `as(...).filter(...)` does the same and `filter` is bread-and-butter idiomatic Rust.

I can understand the readability context a bit but then again, if it's not using the `std` impls then that might imply something non-standard, actually.

I'm fine either way, to be honest!